### DISCUSSION TITLE
Vulnerability tool has detected critical vulnerabilities in Keycloak 20.0.3

### DISCUSSION THREAD

**Discussion by ajaiswal595** [keycloak/keycloak#16517] [Q&A]
### Apache Log4j

Hi, our tool has detected this critical vulnerability does keycloak really vulnerable to these if yes then what is the fix and if no then how can we justify it?

**Location**
![image](https://user-images.githubusercontent.com/8745632/213166407-ae8d65cd-fa36-4aaf-adcc-68bd439ca6a3.png)

Vulnerability:
![image](https://user-images.githubusercontent.com/8745632/213166600-7d116451-6657-4cef-ba92-9a9a8b0a282a.png)

### u\Undertow:
**Location:** 
![image](https://user-images.githubusercontent.com/8745632/213167048-7fd7a4bc-7571-4441-93af-b825cdbd85d3.png)

**Vulnerability:**
![image](https://user-images.githubusercontent.com/8745632/213166920-15b23833-ce12-49f6-a9a7-adf884aae44b.png)

**Comment by ajaiswal595** [comment]
@sschu @pedroigor 

Any thoughts on this?

**Comment by pedroigor** [comment]
As discussed before, we are fixing CVEs from dependencies when upgrading Quarkus. Eventually, we also upgrade dependencies individually.

We are also working on a tool to make it possible to override the server dependencies that might help with this.

@abstractj FYI

**Comment by stianst** [comment]
These are false positives as the issues you are reporting are against log4j, not log4j-jboss-logmanager. Granted the latter does include some classes from Log4j, but doesn't include the affected classes.

**Comment by abstractj** [comment]
@ajaiswal595 we fixed several CVEs between 20.0.3 and 21.0.1. I would suggest updating to the latest release. For the complete list of triaged CVEs that we are aware, please refer to https://github.com/keycloak/keycloak/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Akind%2Fcve

**Reply by darius-m** [reply]
Just as a confirmation, are there any CVE fixes between 20.0.5 and 21.0.1? GHSA-9g98-5mj6-f9mv appears to be the only one that mentions version 21 as being fixed,, but affected versions go up to 20.0.3. Most other advisories mention 20.0.5 as the fix version.

**Reply by abstractj** [reply]
@darius-m the several CVEs I mentioned here are known CVEs detected on third-party dependencies. For Keycloak CVEs and vulnerability fixes we had:

- https://github.com/keycloak/keycloak/security/advisories/GHSA-9g98-5mj6-f9mv
- https://github.com/keycloak/keycloak/security/advisories/GHSA-w354-2f3c-qvg9
- https://github.com/keycloak/keycloak/security/advisories/GHSA-9hhc-pj4w-w5rv
- https://github.com/keycloak/keycloak/security/advisories/GHSA-m4fv-gm5m-4725

**Reply by darius-m** [reply]
Thank you for answering, but my question was a bit of a side - all advisories you have mentioned have gaps between the affected version and the patched version. For example, the first advisory (that I had also linked) mentions that affected versions are `<= 20.0.3`, and patched version is `21.0.1`. However, version 20.0.5 for Keycloak 20 also exists, and is technically in neither category. Other advisories, like GHSA-m4fv-gm5m-4725 have rather big gaps - affected version `< 18.0.0`, and patched version is `20.0.5` - this leaves versions 18, 19, and 20.0.0-20.0.3 basically undefined.

Not sure how these cases should be handled - of course, updating to the latest version is the surest approach, but switching from 20 to 21 may require a bit more work (e.g. because of the removal of the old admin console), and the presence (or lack thereof) of security advisories may be a factor in the decision making process.

**Reply by abstractj** [reply]
@darius-m I see what you're saying and might be worth for us having a separate discussion about it because the topic here is a different subject. But only to bring some clarity, the affected version is set at the moment we confirm the issue, that version is set.

That means that if we have an affected version < 18.0.0, with a patched version 20.0.5, everything in between is vulnerable because the patch was only provided on 20.0.5 release. Of course, that nothing is set in stone, and we can iterate and improve over it. This only the way how it works today.

It should be possible to correct the advisory database, and update with all the affected versions. We can have a separate discussion for that if necessary, so we can explore how to better communicate those issues.

**Reply by abstractj** [reply]
CVEs are updated now to reflect what we discussed @darius-m. Please refer to https://github.com/keycloak/keycloak/security/advisories?state=published
